server/controllers/userAuth.js

```diff
@@ -24,8 +33,102 @@ const sendDemoUser = (req, resp) => {
   resp.redirect('/');
 };
 
+const signUp = (req, resp, next) => {
+  const { email, password, passwordConfirm, username } = req.body;
+  const sanitizedEmail = sanitize(email);
+  const sanitizedPassword = sanitize(password);
+  const sanitizedPasswordConfirm = sanitize(passwordConfirm);
+  const sanitizedUsername = sanitize(username);
+  const errors = {};
+  const { isEmail, isEmpty, isLength, matches } = validator;
+
+  if (!isLength(sanitizedUsername, { min: 1, max: 32 })) {
+    errors.nameError = 'authorization.input.username.invalid';
+  }
+
+  if (isEmpty(sanitizedEmail) || !isEmail(sanitizedEmail)) {
+    errors.emailError = 'authorization.input.email.invalid';
+  }
+
+  if (!matches(sanitizedPassword, /^[^\s]{4,32}$/)) {
+    errors.passwordError = 'authorization.input.password.invalid';
+  }
+
+  if (sanitizedPassword !== sanitizedPasswordConfirm) {
+    errors.passwordConfirmError = 'authorization.input.password.not-match';
+  }
+
+  if (_find(errors, error => error !== undefined)) {
+    return resp.status(406).send(errors);
+  }
+
+  User.findOne({ email: sanitizedEmail })
+    .exec()
+    .then(user => {
+      if (user) {
+        const { _id, displayName, email, idFromProvider, isActive } = user;
+
+        if (!idFromProvider && !isActive) {
+          const signUpHash = crypto.randomBytes(32).toString('hex');
+          const expirationDate = new Date().getTime() + 3600000;
+
+          return User.findOneAndUpdate(
+            { _id },
+            { signUpHash, signUpHashExpirationDate: expirationDate }
+          )
+            .exec()
+            .then(user => {
+              if (!user) {
+                throw new Error();
+              }
+
+              return { displayName, email, signUpHash };
+            });
+        }
+
+        throw new BadRequestException('authorization.user-already-exist');
+      } else {
+        const salt = bcrypt.genSaltSync(12);
+        const hashedPassword = bcrypt.hashSync(password + email, salt);
```

Comment:
you don't need to generate salts with bcrypt (it is part of hash) just use:
`bcrypt.hash(password, 12)` and then `bcrypt.compare(logInData.password, user.password)`